### ISSUE TITLE
Remove use of non-futures aware locks in code executing in the context of tasks

### ISSUE THREAD

**Issue by gterzian** [paritytech/polkadot#180]
We sometimes use locks from `parking_lot` inside futures/streams, and this can potentially block the thread on which the corresponding task is executing, since `parking_lot` is written for use inside threads. However, blocking the thread on the pool in the context of tasks means blocking every other tasks tied to that same thread, being a good example of the kind of blocking that should be avoided inside futures. 

We should look into making such block non-blocking, either through the removal of the lock, which can be achieved by not sharing state between tasks, instead sharing channels, or if a lock is really what we need, we could use something like https://docs.rs/futures/0.1.25/futures/sync/struct.BiLock.html

Further discussion at https://github.com/paritytech/polkadot/pull/149#discussion_r267638366

**Comment by rphmeier** [comment]
I think it's too early to introduce this kind of refactoring. I would wait until the code settles down a bit and we know what we need out of the APIs before we remove the locks. They aren't killing anybody.

**Comment by rphmeier** [comment]
Is it an optimization? Have we profiled and determined that we are bottlenecked on these locks?

**Comment by DemiMarie-parity** [comment]
@rphmeier No, it is a correctness issue.  Using non-futures-aware locks in tasks risks deadlock.

**Comment by bkchr** [comment]
@DemiMarie-parity how?

**Comment by rphmeier** [comment]
I've asked the tokio devs about this at least, and they have not responded that it will interfere with typical wakeup code in any way.

**Comment by bkchr** [comment]
Yeah, that is what I would have expected.

**Comment by gterzian** [comment]
@DemiMarie-parity I'm not sure if using `parking_lot` in the context of tasks/futures increases the risk of deadlock, since deadlock would seem to be more of a logical error that could occur both in the context of threads or tasks. 

What I meant when opening the issue is that `parking_lot`, operates at the level of threads, and isn't aware of any task-based system it might be running inside of. So it's great and very much optimized, but it's optimized for threads, and if the lock is not available, the thread trying to acquire it will be "parked" until the lock is available. 

A task-aware lock, on the other hand, will not park the thread since the thread is supposed to continue servicing many tasks. Instead, a task-aware lock will park the task, and notify it when the lock is available. 

Compare how `parking_lot` unparks a thread waiting on the lock: https://github.com/Amanieu/parking_lot/blob/66d67bbbfe49bb8945940930abfdf5f40ce540a4/core/src/thread_parker/unix.rs#L192

with how `bilock` does it https://github.com/rust-lang-nursery/futures-rs/blob/465ef06d104e2f15098e30971c12642a55a6501b/futures-util/src/lock/bilock.rs#L167

The difference is basically whether the thread can keep servicing other tasks while one tasks is "awaiting" the lock, or whether the thread is blocked waiting on the lock. 

The reason why "tasks/futures should never block", is because blocking a task blocks the thread it is running on, which prevents other tasks from potentially making progress. Instead, tasks should register themselves for being notified when they can make progress, and allow other tasks to be scheduled on the thread in the meantime.

**Comment by bkchr** [comment]
I think we are completely aware how these locks are implemented. As @rphmeier has written in his first post, did you made any performance analysis that proof that the locks are decreasing the performance?

**Comment by gterzian** [comment]
My understanding is that blocking operations are to be avoided in the context of tasks. A contented lock from `parking_lot` will block the thread(past a few optimizations) until it becomes available. My conclusion: non-future aware locks are to be avoided, unless one can guarantee they will never be contented. 

I guess one could indeed run some tests to prove that a lock contention would indeed degrade performance of a series of tasks, and it actually appears self-evident to me. 

Can you explain to me why this would not be the case?

**Comment by gterzian** [comment]
Ok so the answer to "why a contended lock from parking_lot, in the context of tasks which are supposed to not block, somehow wouldn't actually result in degrading of performance?"

seems to be: allocations occurring in for example message-passing could actually be more costly than locking in certain cases, depending on the size of the critical section, see https://github.com/tokio-rs/tokio-core/issues/167#issuecomment-279175083

**Comment by bkchr** [comment]
If you have no concurrent holding of the same lock, you will not see any real performance decrease. 
There is a lot of stuff happening in a multithreaded application and reasoning about certain concepts and their impact on the performance is a rather hard task.

**Comment by gterzian** [comment]
> If you have no concurrent holding of the same lock, you will not see any real performance decrease.

I agree, and also think that's a pretty big "if". 

> reasoning about certain concepts and their impact on the performance is a rather hard task

I agree again, and we probably shouldn't make this a discussion about performance too much. We could instead going forward try to use things like `bilocks`(I haven't used them in the past), maybe also replace shared-state with more use of futures-aware channels, and indeed if you really have to shared a piece of state between tasks and it's too cumbersome to do it with either, have a normal lock and try to keep the critical section as limited as possible. 

There is probably no need to start systematically remove all locks from the futures code, as this issues initially advocated...

**Comment by rphmeier** [comment]
> Can you explain to me why a contended lock from `parking_lot`, in the context of tasks which are supposed to not block, somehow wouldn't actually result in degrading of performance?

This is completely missing the point. Of course it degrades performance somewhat. The question isn't "does it degrade performance at all" but "does it degrade performance in a meaningful way for the product we are trying to build?". 

There isn't any evidence I've seen for the answer being "yes". There are also varying degrees of how much more we can get out of it vs. developer time spent.

**Comment by gterzian** [comment]
I agree. And like I wrote just above, the discussion turned perhaps too much into one about performance. 

Personally, when I think "blocking vs non-blocking", I tend not think in terms of low-level performance, rather in terms of how the business logic "flows", if you see what I mean. 

And another way to look at this, is asking "how do we want to write our business logic?"(incrementally, not perfectly from the start). 

Right now it seems the default is shared-state inside an `Arc`(and a flavor of lock). An alternative is messaging and keeping state inside a task or thread. Yet another is using this future-aware bilock. 

The difference is less about performance, and more about the actual structure of the business logic. 

And if we write a lot of code with arcs and locks, we end up with a different structure of business logic than if we do it with, for example, messages(and the occasional locks when needed, see for example what I did at https://github.com/paritytech/substrate/pull/1857, I am not "anti-locks" in any way :) ). 

It's going to be pretty hard to move the entire system from one style to the next, and performance alone is probably not going to be a strong enough reason to ever actually undertake such a refactoring.

An even more "philosophical" way of thinking about this choice is: our product is a node in a p2p distributed network. Nodes don't share state, instead they send each other messages and eventually reach consensus about the state of the network. When implementing such a node, should we favor shared-state or messaging? Messaging means components inside a node look somewhat similar to the network as a whole. The apparent fractal nature of such a setup points to something good, in my opinion.

**Comment by gterzian** [comment]
> This is completely missing the point. Of course it degrades performance somewhat. The question isn't "does it degrade performance at all" but "does it degrade performance in a meaningful way for the product we are trying to build?". There isn't any evidence I've seen for the answer being "yes". 

I'd like to see if we can bring the discussion away from performance and back on the path of "how do we want to write code". 

If we take locks and shared-state as the baseline, then we indeed need a pretty big argument in terms of performance improvement to refactor to use other means, like bilocks or channels. But why take shared-state as a baseline for how to implement logic?

When I say something like "let's try to make it non-blocking", I don't mean that as a silver bullet for performance, rather it means that the logic we write will be non-blocking. This actually has less to do with using locks vs channels vs something else. 

A good example was the PR [Remove blocking operations in SyncOracle](https://github.com/paritytech/substrate/pull/1852). There I actually replaced the use of channels with simple atomic operations. However the main change was a re-structuring of the logic, going from answering the question "are we major syncing" when it was asked, to instead continuously updating the shared answer. The result is "non-blocking", which probably improved performance but that's only secondary to the structure of the logic becoming non-blocking. 

So next time we have a piece of state that needs to be shared between futures, if we were to try and see if bilocks can be used, or perhaps if state doesn't need to be shared at all and instead messaging can be used, I think we might end-up with a more "non-blocking" kind of logic than if we just default to standard locks. 

That's not an argument for a big refactoring, such changes can be done incrementally over time. I personally it's more enjoyable to at least try to aim for non-blocking logic in concurrent code, and I find it easier to reason about how things run, which helps with making changes over time and in debugging.
